### PR TITLE
chore(docs): document the error types and variants in the `errors` module

### PR DESCRIPTION
**Pull request by iTranscend**
Tracking issue: <https://github.com/near/near-jsonrpc-client-rs/issues/51>

Documented the error types and variants in the `errors` module.

### CODE REVIEW

File: src/errors.rs [hunk 1]
```diff
@@ -1,74 +1,102 @@
...
     TooManyRequests,
+    /// An error returned when the rpc server returns a non-200 status code.
     #[error("the server returned a non-OK (200) status code: [{status}]")]
     Unexpected { status: reqwest::StatusCode },
 }
 
+/// Errors returned by the rpc server.
 #[derive(Debug, Error)]
 pub enum JsonRpcServerError<E> {
+    /// An error returned when the rpc server is unable to validation a client's request.
```
Comment (miraclx) [inline src/errors.rs:76]: This is returned exclusively for Invalid RPC methods. Technically, we should never have this issue with the client (if you stay away from `methods::any`). Still worth documenting, nonetheless.

File: src/errors.rs [hunk 2]
```diff
@@ -1,74 +1,102 @@
...
     #[error("request validation error: [{0:?}]")]
     RequestValidationError(RpcRequestValidationErrorKind),
+    /// An error returned when the server cannot process a request from an rpc client.
```
Comment (miraclx) [inline src/errors.rs:79]: This is an RPC method call error. Like an error returned directly from the method you called.

File: src/errors.rs [hunk 3]
```diff
@@ -1,74 +1,102 @@
...
     #[error("handler error: [{0}]")]
     HandlerError(E),
+    /// An error returned when the rpc server returns an internal server error.
     #[error("internal error: [{info:?}]")]
     InternalError { info: Option<String> },
+    /// An error returned when the rpc server returns a response without context i.e. a response the client doesn't understand.
```
Comment (miraclx) [inline src/errors.rs:85]: +1

understand / expect 🤷🏽

File: src/errors.rs [hunk 4]
```diff
@@ -1,74 +1,102 @@
...
     #[error("error response lacks context: {0}")]
     NonContextualError(RpcError),
+    /// Meta errors returned by the rpc server.
```
Comment (miraclx) [inline src/errors.rs:88]: We should probably be explicit about the fact that these are unexpected status code errors.

File: src/errors.rs [hunk 5]
```diff
@@ -1,74 +1,102 @@
...
     #[error(transparent)]
     ResponseStatusError(JsonRpcServerResponseStatusError),
 }
 
+/// Errors returned by the rpc client.
```
Comment (miraclx) [inline src/errors.rs:93]: Potential errors, maybe?

File: src/errors.rs [hunk 6]
```diff
@@ -1,74 +1,102 @@
...
 #[derive(Debug, Error)]
 pub enum JsonRpcError<E> {
+    /// These errors are returned when there's a problem sending a request to an rpc server.
```
Comment (miraclx) [inline src/errors.rs:96]: Sending / receiving.

This works https://github.com/near/near-jsonrpc-client-rs/pull/84/files#diff-6ff19a1ff611f3835b5a025663cf4dec901ed914626f78847414836439bd1a79R48

File: src/errors.rs [hunk 7]
```diff
@@ -1,74 +1,102 @@
...
     #[error(transparent)]
     TransportError(RpcTransportError),
+    /// These errors are returned whenerver there's a problem with the rpc server.
```
Comment (miraclx) [inline src/errors.rs:99]: Problem not necessarily "with" the server. But at least, returned "from" the server.

**Review by miraclx — CHANGES_REQUESTED** [review]
+1

File: src/errors.rs [hunk 8]
```diff
@@ -6,97 +6,97 @@ use thiserror::Error;
...
-    /// An error returned when the rpc server is unable to validation a client's request.
+    /// An invalid RPC method is called.
     #[error("request validation error: [{0:?}]")]
     RequestValidationError(RpcRequestValidationErrorKind),
-    /// An error returned when the server cannot process a request from an rpc client.
+    /// RPC method call error.
     #[error("handler error: [{0}]")]
     HandlerError(E),
-    /// An error returned when the rpc server returns an internal server error.
+    /// The RPC server returns an internal server error.
```
Comment (miraclx) [inline src/errors.rs:82]:
```suggestion
    /// The RPC server returned an internal server error.
```

File: src/errors.rs [hunk 9]
```diff
@@ -6,97 +6,97 @@ use thiserror::Error;
...
     #[error("internal error: [{info:?}]")]
     InternalError { info: Option<String> },
-    /// An error returned when the rpc server returns a response without context i.e. a response the client doesn't understand.
+    /// The RPC server returns a response without context i.e. a response the client doesn't expect.
```
Comment (miraclx) [inline src/errors.rs:85]: ditto.

```suggestion
    /// The RPC server returned a response without context i.e. a response the client doesn't expect.
```

**Review by miraclx — CHANGES_REQUESTED** [review]

File: src/errors.rs (hunk 1 above)
Comment (miraclx) [reply src/errors.rs]: Okay, so this isn't just for invalid RPC methods. But also for cases where the RPC method is unable to parse the provided arguments.

**Review by miraclx — CHANGES_REQUESTED** [review]

File: src/errors.rs (hunk 1 above)
Comment (miraclx) [reply src/errors.rs]: Or if the payload sent to the server, while a valid JSON RPC message, isn't structured as a request.

The inverse of https://github.com/near/near-jsonrpc-client-rs/blob/091fb5385d0212716f343e0cbf7752ea1ce210c4/src/errors.rs#L34-L36

**Comment by iTranscend** [comment]
Requested changes addressed ✅

**Comment by frol** [comment]
I will merge all the doc PRs as is, and we will address any further issues in separate PRs going forward.

**Pull request merged**